packages/amplify-category-function/src/provider-utils/awscloudformation/utils/environmentVariablesHelper.ts

```diff
@@ -322,3 +313,62 @@ const getStoredKeyValue = (resourceName: string, envName?: string): Record<strin
 const setStoredKeyValue = (resourceName: string, newKeyValue: $TSAny, envName?: string): void => {
   getEnvParamManager(envName).getResourceParamManager(categoryName, resourceName).setAllParams(newKeyValue);
 };
+
+type MissingEnvVarsConfig = {
+  funcName: string;
+  missingEnvVars: {
+    environmentVariableName: string;
+    cloudFormationParameterName: string;
+  }[];
+}[];
+
+const getMissingEnvVarsError = (
+  missingVars: MissingEnvVarsConfig,
+  appId: string | undefined,
+  envName: string | undefined,
+): AmplifyError => {
+  const message = `This environment is missing some function environment variable values.`;
+  const missingEnvVarsDetails = missingVars
+    .map(({ missingEnvVars, funcName }) => {
+      const missingEnvVarsString = missingEnvVars.map((missing) => missing.environmentVariableName).join(', ');
+      return `Function ${funcName} is missing values for environment variables: ${missingEnvVarsString}`;
+    })
+    .join('\n');
+  if (appId === undefined) {
+    return new AmplifyError('EnvironmentConfigurationError', {
+      message,
+      details: missingEnvVarsDetails,
+      resolution: `An AppId could not be determined. Make sure your project is initialized and rerun 'amplify push' without '--yes' to fix.`,
+    });
+  }
+
+  if (envName === undefined) {
+    return new AmplifyError('EnvironmentConfigurationError', {
+      message,
+      details: missingEnvVarsDetails,
+      resolution: `A current environment name could not be determined for fetching missing parameters. Make sure your project is initialized using "amplify init"`,
+    });
+  }
```

Comment:
Do we not want to distinguish between these two error types in telemetry? `resolution` is not part of it so we won't be able to tell if envName was `missing` or `appId`

Reply:
We have `InvalidAmplifyAppIdError` and `EnvironmentNameError` that I could use. However, it looks like existing usage of `EnvironmentNameError` is for malformed env names and `InvalidAmplifyAppIdError` is for when an appId specified in input params conflicts with a local appId. I could make new error types for each of these cases but that seems like overkill. We would have the stack trace in the error data right?

Reply:
We do have the stack trace but it's not as useful. I was thinking maybe appending or prepending `envName not found` to the message field.